### ISSUE TITLE
DicomServer does not response when running as Windows service

### ISSUE THREAD

**Issue by mrmokilok** [fo-dicom/fo-dicom#1039]
### Expected behavior
DicomServer instance runs as Windows service and responds to Echo request. Verified with a statistic events from running, received request and successful Echo.

### Actual behavior
Windows shows running service, but no statistics events are stored.

### Steps to reproduce the behavior
Application frame:
Program.cs
`public class Program
    {
        public static void Main(string[] args)
        {
            CreateHostBuilder(args).Build().Run();

        }
        public static IHostBuilder CreateHostBuilder(string[] args)
        {
            string environment = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT");

            var builder = new ConfigurationBuilder()
                .SetBasePath(Directory.GetCurrentDirectory())
                .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
                .AddJsonFile($"appsettings.{environment}.json", optional: true);

            IConfigurationRoot configuration = builder.Build();

            return Host.CreateDefaultBuilder(args)
                .ConfigureServices((hostContext, services) =>
                {
                    services.AddSingleton<IConfigurationRoot>(configuration);
                    services.AddHostedService<Worker>();
                }).UseWindowsService();
        }
    }`

Worker.cs:
`public class Worker : BackgroundService
    {
        private readonly ILogger<Worker> _logger;

        public IConfiguration _configuration;

        public Worker(ILogger<Worker> logger, IConfiguration configuration)
        {
            _logger = logger;
            _configuration = configuration;
        }

        /// <summary>
        /// Worker execute creates DicomServer instance, which waits return key to end. 
        /// Server is recreated after the key if cancellation token (stoppingToken) isn't provided.
        /// </summary>
        /// <param name="stoppingToken"></param>
        /// <returns></returns>
        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            while (!stoppingToken.IsCancellationRequested)
            {
                _logger.LogInformation("Worker running at: {time}", DateTimeOffset.Now);
                // preload dictionary to prevent timeouts
                var dict = DicomDictionary.Default;

                // start DICOM server on port from command line argument or 11112
                int port = int.Parse(_configuration["DicomListenerPort"]);
                string ipAddress = _configuration["DicomListenerIP"];
                Console.WriteLine($"Starting C-Store SCP server on port {port}");

                try
                {
                    using (var server = DicomServer.Create<CStoreSCP>(ipAddress, port))
                    {
                        await ImageStudyDbHelper.AddStatisticEventAsync(_configuration.GetConnectionString("ImageStudyDBConnectionString"), (int)EventTypes.Listener_Started, "Listener on port: " + port);

                        // end process
                        //Console.WriteLine("Press <return> to end...");
                        //Console.ReadLine();
                        while (!stoppingToken.IsCancellationRequested)
                        {
                            Thread.Sleep(500);
                        }
                    }
                }
                catch(Exception e)
                {
                    Console.WriteLine("Failed to run DicomServer: ");
                    Console.Write(e);
                }
            }
        }
    }`

Run command prompt as an Administrator and:
1) Install service with: sc create TestService BinPath=C:\full\path\to\publish\dir\WindowsServiceExample.exe

2) Run service with: sc start TestService

### fo-dicom version and OS/platform
v4.0.4
Windows 10
netcoreapp3.1

**Comment by gofal** [comment]
Thats interesting. I assume the issue is in how the WindowsService work.

There are 2 interfaces:
- `IHostedService` with the methods `.StartAsync` and `.StopAsync`
- `BackgroundService` with the abstract method `.ExecuteAsync`
The class `BackgroundService` implements the interface `IHostedService`, which is required by the extension method `services.AddHostedService<T>`.

See here: 
https://docs.microsoft.com/en-us/dotnet/architecture/microservices/multi-container-microservice-net-applications/background-tasks-with-ihostedservice#implementing-ihostedservice-with-a-custom-hosted-service-class-deriving-from-the-backgroundservice-base-class
The default implementation of `BackgroundService` calls `ExecuteAsync`, but does not await this, but stores the Task in a class field and returns Task.completed. 
So I assume, the `StartAsync`, which then calls `ExecuteAsync` has to complete when used as WindowService and should not contain an endless loop.

**Comment by gofal** [comment]
This is, how my implementation looks like, which works without problems:

    public class StoreService : BackgroundService
    {

        private readonly ILogger<StoreService> _logger;
        // .. some other configuration fields etc..

        private IDicomServer _server;

        public StoreService(ILogger<StoreService> logger, IOptions<ServiceSettings> settings /* more params ... */)
        {
            _logger = logger;
            _settings = settings?.Value ?? throw new ArgumentNullException(nameof(settings));
            // ....
        }


        // this method has to be implemented because it is abstract, even if it does nothing
        protected override Task ExecuteAsync(CancellationToken stoppingToken)
        {
            return Task.CompletedTask;
        }


        public override Task StartAsync(CancellationToken cancellationToken)
        {
            _logger.LogInformation($"StoreService starting on port {_settings.LocalPort}");

            // preload dictionary to prevent timeouts
            _ = DicomDictionary.Default;

            // doing initial setups, like setting the transcoder manager etc..
            TranscoderManager.SetImplementation(new NativeTranscoderManager());

           // now the server is created and stored in a class field. No loop etc.
            _server = DicomServer.Create<StoreSCP>(
                _settings.LocalPort, 
                new StoreScpOptions { Settings = _settings, ServiceProvider = _serviceProvider },
                certificateName: null,
                options: new DicomServiceOptions { LogDimseDatasets = _settings.DicomLogging },
                fallbackEncoding: null,
                logger: new StoreLogger(_logger, _settings)
            );

            return base.StartAsync(cancellationToken);
        }


        public override Task StopAsync(CancellationToken cancellationToken)
        {
           // when StopAsync is called, the server is stopped, if it was running
            _logger.LogInformation("StoreService stopping");
            if (_server != null)
            {
                _server.Stop();
            }

            return base.StopAsync(cancellationToken);
        }


        public override void Dispose()
        {
            _server?.Dispose();
            base.Dispose();
        }

    }

**Comment by mrmokilok** [comment]
That really is interesting! I only implemented the ExecuteAsync, as I understood from other documentation that it's the only one needed. The link to MS docs shows that the StartAsync on base calls it.

I tried adapting your example, but I think I need the implementation to ServiceSettings and NativeTransCoderManager to get it working. My adaptation worked same as my previous code, it ran as console app but didn't respond to Echo as a service.

I test Echo with Charruasoft TestSCU (http://www.charruasoft.com/products/free/index.php).

Here's my adapted code:

`public class StoreService : BackgroundService
    {

        private readonly ILogger<StoreService> _logger;
        public IConfiguration _configuration;

        private IDicomServer _server;
        private Guid _serverRunGuid;

        public StoreService(ILogger<StoreService> logger, IConfiguration configuration)
        {
            _logger = logger;
            _configuration = configuration;
            _serverRunGuid = Guid.NewGuid();
            // ....
        }


        // this method has to be implemented because it is abstract, even if it does nothing
        protected override Task ExecuteAsync(CancellationToken stoppingToken)
        {
            return Task.CompletedTask;
        }


        public override Task StartAsync(CancellationToken cancellationToken)
        {
            _logger.LogInformation("Worker running at: {time}", DateTimeOffset.Now);
            // preload dictionary to prevent timeouts
            var dict = DicomDictionary.Default;

            // start DICOM server on port from command line argument or 11112
            int port = int.Parse(_configuration["DicomListenerPort"]);
            string ipAddress = _configuration["DicomListenerIP"];
            Console.WriteLine($"Starting C-Store SCP server on port {port}");
            

            // doing initial setups, like setting the transcoder manager etc..
            //TranscoderManager.SetImplementation(new NativeTranscoderManager());

            // now the server is created and stored in a class field. No loop etc.
            try
            {
                _server = DicomServer.Create<CStoreSCP>(ipAddress, port);
                ImageStudyDbHelper.AddStatisticEventAsync(_configuration.GetConnectionString("ImageStudyDBConnectionString"), (int)EventTypes.Listener_Started, "Listener on port: " + port, guid: _serverRunGuid).Wait();
            }
            catch (Exception e)
            {
                ImageStudyDbHelper.AddStatisticEventAsync(_configuration.GetConnectionString("ImageStudyDBConnectionString"), (int)EventTypes.Listener_Start_Failed, "Listener on port: " + port, JsonConvert.SerializeObject(e), _serverRunGuid).Wait();
                Console.WriteLine("Failed to run DicomServer: ");
                Console.Write(e);
            }

            return base.StartAsync(cancellationToken);
        }


        public override Task StopAsync(CancellationToken cancellationToken)
        {
            // when StopAsync is called, the server is stopped, if it was running
            _logger.LogInformation("StoreService stopping");
            ImageStudyDbHelper.AddStatisticEventAsync(_configuration.GetConnectionString("ImageStudyDBConnectionString"), (int)EventTypes.Listener_Abort, "Listener aborted", guid: _serverRunGuid).Wait();
            if (_server != null)
            {
                _server.Stop();
            }

            return base.StopAsync(cancellationToken);
        }


        public override void Dispose()
        {
            _server?.Dispose();
            base.Dispose();
        }

    }`

**Comment by MadamLaunch** [comment]
Where are you guys getting `StoreSCP/CStoreSCP` from? I'm trying to create a server myself, and there's no record of that variable/class/enumerator existing anywhere within fo-dicom.

**Comment by mrmokilok** [comment]
You can find example here:
https://github.com/fo-dicom/fo-dicom-samples/tree/master/Desktop/C-Store%20SCP
in
Program.cs

Note that best place to ask questions is the fo-dicom gitter:
https://gitter.im/fo-dicom/fo-dicom

ke 3. kesäk. 2020 klo 19.50 MadamLaunch (notifications@github.com)
kirjoitti:

**Comment by gofal** [comment]
I now again created a complete fresch applicateion copying the code from your issue-description. I ran from visual studio and tested verification with TestScu.exe from CharruaSoft. 
Then installed the exe as service with "sc create ...", started the service and again tested with TestScu.exe: it worked!

I assume some firewall-issue on your machine. Did you check that?
install Telnet client on your machine. You should at least be able to open a telnet connection to the ip/port. of course the DicomServer then will throw exceptions because it cannot read from the stream created. But this will show you if you are facing an application-problem or a network-problem.

**Comment by mrmokilok** [comment]
Tested with telnet, but of course it doesn't answer if the port listener doesn't start. 

I actually tested bit further and moved init log event write before starting the Fo-dicom server and it doesn't get written. So, the StartAsync method doesn't get called even though the Task Manager shows that the service is running. Can you share your working solution via Github?

**Comment by gofal** [comment]
[ConsoleApp2.zip](https://github.com/fo-dicom/fo-dicom/files/4744891/ConsoleApp2.zip)
This is the solution I mentioned above.

**Comment by mrmokilok** [comment]
I'm back, working on this again and I am puzzled! Your example app answers to echo, just fine but I can't get mine to do it when running as a service. Again, in console it works.

@gofal, you mentioned that you had challenges and comparing the project file in your example to VS Worker template, it doesn't list Microsoft.NET.Sdk.Worker as Sdk but Microsoft.NET.Sdk and the OutputType exe is set there. Do these have anything to do with the challenges you mentioned?

**Comment by gofal** [comment]
no, the challanages so far I know about are that the livetime of the DicomServer like you have to hold the instance of the dicomserver as class field, or it are about finding the codec dll in the same dir when starting as service or such.
But for your special problem I have no experience. Sorry.

**Comment by mrmokilok** [comment]
I'm sorry, could you rephrase that, I didn't quite understand? If one of the challenges was keeping the listener alive, I did have few try outs where the service stopped automatically after few seemingly insignificant changes.

**Comment by gofal** [comment]
Clsoing the issue due to inactivity.
If with the latest version of fo-dicom there is still an issue, please reopen this issue and provide actual informations.
